### PR TITLE
Add PostAggregators to generator cache keys for top-n queries

### LINKED ISSUE #3719: Improper topN caching when sorting by postaggregator
When a topN is sorting by a postaggregator, the postaggregator name is part of the query cache key but the definition is not. But, the postaggregator should be part of the cache key too, since it's used to determine which result rows to keep for a particular segment in the cache. If the postaggregator is redefined but keeps the same name, the query improperly uses previously cached results.

I think fixing this for real will require adding a getCacheKey method to PostAggregators.

If we want to hold off on that until 0.10.0 then we could work around it in the meantime by disabling caching for any topN query that sorts on a postaggregator. But this has a performance cost that could hit more people than the bug does.

### PR DESCRIPTION
**Pull request by jihoonson**
This PR is to fix #3719.

This change is [Reviewable](https://reviewable.io/reviews/druid-io/druid/3899)

### CODE REVIEW

File: processing/src/test/java/io/druid/query/cache/CacheKeyBuilderTest.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(Parameterized.class)
+public class CacheKeyBuilderTest
```
Comment (drcrallen) [inline processing/src/test/java/io/druid/query/cache/CacheKeyBuilderTest.java]: Can you add a test for the following:

Item 1:
String1 : `a`
String2 : `b`

Item 2:
String 1:`ab`

Comment (jihoonson) [inline processing/src/test/java/io/druid/query/cache/CacheKeyBuilderTest.java]: I added tests. Thanks.

File: processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+  public CacheKeyBuilder appendCacheable(@Nullable Cacheable input)
+  {
+    appendItem(input == null ? EMPTY_BYTES : input.getCacheKey());
+    return this;
+  }
+
+  public CacheKeyBuilder appendCacheableList(List<? extends Cacheable> inputs)
+  {
+    for (Cacheable input : inputs) {
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java]: can the list be null in any case? We might as well just add null a check here.

File: processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+  public CacheKeyBuilder appendString(@Nullable String input)
+  {
+    appendItem(stringToByteArray(input));
+    return this;
+  }
+
+  public CacheKeyBuilder appendStringList(List<String> input)
+  {
+    for (String eachStr : input) {
+      appendItem(stringToByteArray(eachStr));
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java]: Ok, I'll be more explicit here.

If you call 

```java
appendStringList(ImmutalbeList.of("a","b"));
```

in one cache key builder, and 

```java
appendStringList(ImmutableList.of("ab"))
```

in another, they will both evaluate (incorrectly) to the same cache key, if I'm reading the code correctly. All strings should be terminated with a byte that is NOT a valid UTF-8 character start (like `0xFF`)

**Comment by fjy** [comment]
👍

File: server/src/test/java/io/druid/client/CachingClusteredClientTest.java [hunk 4]
```diff
@@ -205,7 +205,7 @@
       )
   );
   private static final List<AggregatorFactory> RENAMED_AGGS = Arrays.asList(
-      new CountAggregatorFactory("rows2"),
+      new CountAggregatorFactory("rows"),
```
Comment (gianm) [inline server/src/test/java/io/druid/client/CachingClusteredClientTest.java]: What's the idea behind this change, and the makeTopNResults -> makeTopNResultsWithoutRename change? It's not clear to me why the new test is better.

File: processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java (hunk 2 above)
Comment (jihoonson) [inline processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java]: I don't see any case now, but it will be helpful for future uses. I'll add it.

File: processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java (hunk 3 above)
Comment (jihoonson) [inline processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java]: @drcrallen, thanks. I'll add tests.

@gianm, thank you for the good suggestion. I'm concerned with the cache key size. How about adding some keys for each type like below?

```java
static final byte BYTE_KEY = 0;
static final byte BYTE_ARRAY_KEY = 1;
...
static final byte INT_KEY = 10;
static final byte INT_ARRAY_KEY = 11;
...

public CacheKeyBuilder appendByte(byte input)
{
  appendItem(BYTE_KEY, byteToByteArray(input));
  return this;
}

public CacheKeyBuilder appendByteArray(byte[] input)
{
  appendItem(BYTE_ARRAY_KEY, input);
  return this;
}

private void appendItem(byte key, byte[] bytes)
{
  ...
}
...
```

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryQueryToolChest.java [hunk 5]
```diff
@@ -356,45 +355,13 @@ public String apply(DimensionSpec input)
...
-        return buffer
-            .put(havingBytes)
-            .put(limitBytes)
-            .array();
+        return new CacheKeyBuilder(GROUPBY_QUERY, CacheKeyBuilder.EMPTY_BYTES)
+            .appendByte(CACHE_STRATEGY_VERSION)
+            .appendCacheable(query.getGranularity())
+            .appendCacheable(query.getDimFilter())
+            .appendCacheableList(query.getAggregatorSpecs())
+            .appendCacheableList(query.getDimensions())
```
Comment (jihoonson) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryQueryToolChest.java]: Thanks. It's my bad. I'll fix it.

After #3873, the group by result can be different if the limit-push-down is enabled and the ordering keys contain non-dimensions. I think this should also be considered to make the cache key.

File: processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java (hunk 3 above)
Comment (gianm) [inline processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java]: @jihoonson, that sounds like a good way to save some space.

File: server/src/test/java/io/druid/client/CachingClusteredClientTest.java (hunk 4 above)
Comment (jihoonson) [inline server/src/test/java/io/druid/client/CachingClusteredClientTest.java]: This was not for a better test, but to make unit tests passed.

```CachingClusteredClientTest.testTopN*``` tests check the cached result, and here the same post aggregator should be used except their names. So, I just changed test queries to use ```POST_AGG``` which is also used for caching like [here](https://github.com/druid-io/druid/pull/3899/files/ddb4eee8859b1bd5d433b2df92e742e2a2ee9dc0#diff-7254bc82bb1d7b6afc78e03268c96806R821). 

To make ```POST_AGG``` work properly, the ```FieldAccessPostAggregator.fieldName``` should be the one specified in ```AGGS```. So, I changed the name of ```CountAggregatorFactory``` from ```rows2``` to ```rows```.

Since there is another renamed field ```impers2``` in ```RENAMED_AGG```, I thought it would not be a problem in other tests for group-by with renaming. If not, please let me know.

The change in makeTopNResultsWithoutRename() is just for my convenience. If its name is verbose, I'll change it.

Comment (gianm) [inline server/src/test/java/io/druid/client/CachingClusteredClientTest.java]: Ah I see what's going on. There's nothing wrong with verbose methods (usually I prefer them) so that's fine.

File: processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java [hunk 6]
```diff
@@ -306,27 +303,15 @@ public String apply(@Nullable AggregatorFactory input)
...
-            .put(aggregatorBytes)
-            .array();
+        return new CacheKeyBuilder(TOPN_QUERY, CacheKeyBuilder.EMPTY_BYTES)
+            .appendCacheable(query.getDimensionSpec())
+            .appendCacheable(query.getTopNMetricSpec())
+            .appendInt(query.getThreshold())
+            .appendCacheable(query.getGranularity())
+            .appendCacheable(query.getDimensionsFilter())
+            .appendCacheableList(query.getAggregatorSpecs())
+            .appendCacheableList(query.getPostAggregatorSpecs())
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java]: Until now, just changing the post-aggregators on a topN (maybe adding new arithmetic, something like that) would still allow the cache to be used. This is a bug if we're sorting by the post-aggregator but is fine otherwise.

What do you think about changing the cache key generation so it only includes post-aggs we're sorting on (and any transitive dependencies based on `getDependentFields`)?

Comment (jihoonson) [inline processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java]: Sounds good. I'll do.

File: processing/src/main/java/io/druid/query/aggregation/post/PostAggregatorIds.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  public static final byte DOUBLE_LEAST = 5;
+  public static final byte EQUAL_BUCKETS = 6;
+  public static final byte EXPRESSION = 7;
+  public static final byte FIELD_ACCESS = 8;
+  public static final byte HYPER_UNIQUE_FINALIZING = 9;
+  public static final byte JAVA_SCRIPT = 10;
+  public static final byte LONG_GREATEST = 11;
+  public static final byte LONG_LEAST = 12;
+  public static final byte MAX = 13;
+  public static final byte MIN = 14;
```
Comment (jihoonson) [inline processing/src/main/java/io/druid/query/aggregation/post/PostAggregatorIds.java]: Changed.

File: processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java (hunk 3 above)
Comment (jihoonson) [inline processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java]: I realized that both type keys and list sizes are needed to avoid this problem. I added them to CacheKeyBuilder.

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryQueryToolChest.java (hunk 5 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryQueryToolChest.java]: The limit push down in #3873 is applied when merging the per-segment results, so taking the limit out of the cache key should be fine there as well

**Comment by jihoonson** [comment]
@drcrallen, @gianm thanks for your review. I addressed comments.
Additionally, I added the below feature to CacheKeyBuilder.
* When a collection is appended to CacheKeyBuilder, that collection is sorted by its items' byte representation. This is to guarantee that collections of same items but in different orders make the same result

File: processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java [hunk 8]
```diff
@@ -291,6 +291,76 @@ public String apply(@Nullable AggregatorFactory input)
     return TYPE_REFERENCE;
   }
 
+  @VisibleForTesting
+  static Collection<PostAggregator> findPostAggregatorsForSort(
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java]: This looks redundant to the already existing `private static List<PostAggregator> prunePostAggregators(TopNQuery query)` which calls `AggregatorUtil.pruneDependentPostAgg` (used for similar functionality in other parts of the topN system)

**Comment by gianm** [comment]
@jihoonson, thx for the update. had a couple further comments on the new stuff.

File: processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java (hunk 8 above)
Comment (jihoonson) [inline processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java]: Thanks. I changed to use that method.

File: processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+    if (collection.size() > 0) {
+      List<byte[]> byteArrayList = Lists.newArrayListWithCapacity(collection.size());
+      int totalByteLength = 0;
+      for (T eachItem : collection) {
+        final byte[] byteArray = serializeFunction.apply(eachItem);
+        totalByteLength += byteArray.length;
+        byteArrayList.add(byteArray);
+      }
+
+      // Sort the byte array list to guarantee that collections of same items but in different orders make the same result
```
Comment (jihoonson) [inline processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java]: Good catch. Thanks. I added appendCacheablesIgnoringOrder(). 
I checked which one of appendCacheables and appendCacheablesIgnoringOrder should be called when building cache keys, but am not sure for SketchSetPostAggregator. Please verify it.

**Comment by jihoonson** [comment]
@gianm, thanks. I addressed your comments. 

I think we need more tests to verify query caching with various types of fields, aggregators, post aggregators, etc. To make the tests concrete, I think it would be better to add more classes which are specialized for each query type. These tests are parameterized with original query, renamed parameters, and parameters of different order. However, this will cause a huge change in tests. I'm not sure about adding these tests in this PR. Maybe better to add in a following PR? What do you think?

**Comment by gianm** [comment]
That test sounds great -- we don't have amazing coverage of caching and cache key right now. I think it makes sense to do it in a followup PR.

File: extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchSetPostAggregator.java [hunk 10]
```diff
@@ -165,10 +165,29 @@ public int hashCode()
...
+    if (preserveFieldOrder(func)) {
+      builder.appendCacheables(fields);
+    } else {
+      builder.appendCacheablesIgnoringOrder(fields);
+    }
+
+    return builder.build();
+  }
+
+  private static boolean preserveFieldOrder(SketchHolder.Func func)
```
Comment (gianm) [inline extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchSetPostAggregator.java]: Suggest calling this `preserveFieldOrderInCacheKey` for better clarity.

File: processing/src/main/java/io/druid/query/aggregation/post/ArithmeticPostAggregator.java [hunk 11]
```diff
@@ -163,6 +169,21 @@ public String toString()
            '}';
   }
 
+  private static boolean preserveFieldOrder(Ops op)
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/aggregation/post/ArithmeticPostAggregator.java]: Similar, suggest calling this `preserveFieldOrderInCacheKey`.

File: processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java [hunk 12]
```diff
@@ -100,7 +97,7 @@ public String apply(@Nullable AggregatorFactory input)
     ).toArray(new String[0]);
   }
 
-  private static List<PostAggregator> prunePostAggregators(TopNQuery query)
+  public static List<PostAggregator> prunePostAggregators(TopNQuery query)
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java]: This doesn't need to be public, it's only called from within this file.

**Comment by jihoonson** [comment]
@gianm, fixed. Thanks!

File: processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+
+public class CacheKeyBuilder
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java]: Can you add docs here for a high level overview of how this class works?

File: processing/src/main/java/io/druid/query/groupby/having/HavingSpec.java [hunk 14]
```diff
@@ -43,20 +44,20 @@
...
     @JsonSubTypes.Type(name = "filter", value = DimFilterHavingSpec.class)
 })
-public interface HavingSpec
+public interface HavingSpec extends Cacheable
 {
   // Atoms for easy combination, but for now they are mostly useful
   // for testing.
-  public static final HavingSpec NEVER = new NeverHavingSpec();
-  public static final HavingSpec ALWAYS = new AlwaysHavingSpec();
+  HavingSpec NEVER = new NeverHavingSpec();
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/groupby/having/HavingSpec.java]: why this change?

File: processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java [hunk 15]
```diff
@@ -306,27 +305,22 @@ public String apply(@Nullable AggregatorFactory input)
...
-            .array();
+        final CacheKeyBuilder builder = new CacheKeyBuilder(TOPN_QUERY)
+            .appendCacheable(query.getDimensionSpec())
+            .appendCacheable(query.getTopNMetricSpec())
+            .appendInt(query.getThreshold())
+            .appendCacheable(query.getGranularity())
+            .appendCacheable(query.getDimensionsFilter())
+            .appendCacheablesIgnoringOrder(query.getAggregatorSpecs());
+
+        final List<PostAggregator> postAggregators = prunePostAggregators(query);
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java]: Does this mean that if I change post aggregators that are not part of the sorted that I will always get a cached result?

File: processing/src/main/java/io/druid/query/groupby/having/HavingSpec.java (hunk 14 above)
Comment (gianm) [inline processing/src/main/java/io/druid/query/groupby/having/HavingSpec.java]: They don't do anything for interface fields.

File: processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java (hunk 15 above)
Comment (gianm) [inline processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java]: That's the idea: only postaggs involved in the sorting could affect the topN by-segment results, so they're the only ones that need to be in the cache key.

File: processing/src/main/java/io/druid/query/groupby/having/HavingSpec.java (hunk 14 above)
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/groupby/having/HavingSpec.java]: gotcha

File: processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java (hunk 15 above)
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java]: oh, duh, because the post-aggs can be calculated from the cached results.

Are they? My mind is drawing a blank there. Do post-aggs get calculated AFTER cached results?

Comment (gianm) [inline processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java]: Yes, they are calculated after pulling from cache. But in this case, a postagg can affect the sort order of cached results, even if it's not actually _in_ the cached results. So it still needs to be part of the topN cache key (to fix #3719).

Comment (drcrallen) [inline processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java]: Had to go look myself to make sure, `io.druid.client.CachingClusteredClient#mergeCachedAndUncachedSequences` and `io.druid.client.CachingQueryRunner#run` seem to be able to handle it but holy cow, I forgot how convoluted the data paths are for actually running a query.

**Comment by drcrallen** [comment]
Requesting more class docs for the cache helper, but not strong enough to block the PR

**Comment by gianm** [comment]
Going to merge this one then, but @jihoonson could you please do a follow up adding the javadocs that @drcrallen mentioned to CacheKeyBuilder? thx

**Comment by jihoonson** [comment]
@gianm, @drcrallen thanks for your review. I opened #3933 for adding docs.
